### PR TITLE
Adds post about client-side caching

### PR DESCRIPTION
**Pull request by itamarhaber**

### CODE REVIEW

File: content/posts/redis-assisted-client-side-caching-in-python.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+## Design notes
+I was leaning towards prototyping this in Python, and a short [informal poll](https://twitter.com/itamarhaber/status/1207773964648550403) supported this approach. I had a three-part setup in mind:
+
+1. A connection to Redis to read data
+1. A cache to keep the data local
+1. A manager to tie everything nicely together
+
+To make the connection, I chose [redis-py](https://github.com/andymccurdy/redis-py). It provides the `Redis` class that is a straight-forward zero-fuss client, and Python's nature makes extending it easy.
+
+The requirements from the cache component are basic, so I was perfectly happy adapting the [LRU cache example in Python's `OrederedDict` documentation](https://docs.python.org/3/library/collections.html#ordereddict-examples-and-recipes).
```
Comment (gavrie) [inline content/posts/redis-assisted-client-side-caching-in-python.md]: Typo: `OrderedDict`

File: content/posts/redis-assisted-client-side-caching-in-python.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+This reduces the server resources required to track multiple keys for multiple clients. The invalidation messages sent by Redis, therefore, consist of _slots_ that need to be invalidated rather than key names. It is up to the client to infer the relevant key names that need to be removed from its cache given the slot.
+
+That means the client needs to employ the same hashing function to track how the keys in the local cache map to slots. That lets us perform slot-based invalidation of the client's cache when an invalidation notification arrives. For that, we'll use the `add()` and `discard()` methods when keys are added and discarded from the local cache, respectively.
+
+```python
+    def slot(key):
+        ''' Returns the slot for a key '''
+        crc = crc64(key)
+        crc %= 16777216
```
Comment (gavrie) [inline content/posts/redis-assisted-client-side-caching-in-python.md]: Writing `0x1000000` would be more readable.

File: content/posts/redis-assisted-client-side-caching-in-python.md [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        self.slots[slot].add(key)
+
+    def discard(self, key):
+        ''' Removes a key from the internal tracking table '''
+        slot = self.slot(key)
+        self.slots[slot].discard(key)
+```
+
+# Handling invalidation
+How an invalidation message is sent to a tracked client depends on the [Redis Serialization Protocol (RESP)](https://redis.io/topics/protocol) that the client is using. Earlier versions of Redis use RESP2, but its successor [RESP3](https://github.com/antirez/RESP3/blob/master/spec.md) is already present in the Redis 6 and will deprecate the older protocol completely in Redis 7.
```
Comment (gavrie) [inline content/posts/redis-assisted-client-side-caching-in-python.md]: `s/the Redis 6/Redis 6/`

File: content/posts/redis-assisted-client-side-caching-in-python.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    def get_connection(self, *args, **kwargs):
+        ''' Returns a cached Redis connection '''
+        conn = CachedRedis(self, connection_pool=self.pool, *args, **kwargs)
+        return conn
+```
+
+## Some measurements
+This post isn't about benchmarking or Python's performance per se, but it’s important to understand the mechanism's impact. For that purpose, I've used the [benchmark.py](https://github.com/itamarhaber/rsacsc-py/blob/master/benchmark.py) script on a 2019 MacBook Pro with a Redis instance running locally using defaults (except that I turned off snapshotting).
+
+Before performing the tests, the benchmark script populates the database with 1.000 keys and sets up the cache manager with a capacity for 100. It then runs several timed tests to measure performance. Each test is repeated five times for both types of connections: regular and cached.
```
Comment (gavrie) [inline content/posts/redis-assisted-client-side-caching-in-python.md]: `s/1.000/1000/` or `1,000`

File: content/posts/redis-assisted-client-side-caching-in-python.md [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+However, repeating the test on a dataset that fits into the cache—that is, only 100 keys—shows more encouraging results. While the first cached run shows an increase in latency, subsequent ones reduce latency by two orders of magnitude:
+
+<div id="chart_2"></div>
+
+The next test is named `eleven_reads` because it reads every key in the database once along with 10 other keys that are always the same ones. This highly synthetic use case provides an even more dramatic proof of the cache's benefit (despite that not being the purpose, per se).
+
+<div id="chart_3"></div>
+
+The last test extends `eleven_reads` with an additional write request to one of the 10 constant keys, which triggers the invalidation of a part of the cache. Latency of cached runs increase slightly, both because of the extra write command but also due to the need to refetch the contents of the cache:
```
Comment (gavrie) [inline content/posts/redis-assisted-client-side-caching-in-python.md]: "Latency of cached runs increase slightly" => "...increases slightly"

File: content/posts/redis-assisted-client-side-caching-in-python.md [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+The move in RESP3 from PubSub broadcasting to connection-specific notifications also means that clients will get invalidation notifications only for slots that interest them. That means less resources spent on communication, and less is better.
+
+Regardless of the RESP version used, client authors can use RSACSC for caching much more than just the GETting of entire strings. The mechanism is agnostic of the actual data structure used for storing the key's value, so all core Redis types and any custom ones declared by modules can be used with it.
+
+Furthermore, instead of just caching key-value tuples, the client can cache requests and their replies (while keeping track of the keys involved). Doing so enables caching of substrings returned [`GETRANGE`](https://redis.io/commands/getrange), list elements obtained via [`LRANGE`](https://redis.io/commands/lrange), or virtually any other type of query.
```
Comment (gavrie) [inline content/posts/redis-assisted-client-side-caching-in-python.md]: "substrings returned `GETRANGE`" => "...returned by..."

File: content/posts/redis-assisted-client-side-caching-in-python.md (hunk 2 above)
Comment (itamarhaber) [reply content/posts/redis-assisted-client-side-caching-in-python.md]: Right. I'll do `&= 0xffffff` instead though.

File: content/posts/redis-assisted-client-side-caching-in-python.md (hunk 1 above)
Comment (itamarhaber) [reply content/posts/redis-assisted-client-side-caching-in-python.md]: thanks.

File: content/posts/redis-assisted-client-side-caching-in-python.md (hunk 3 above)
Comment (itamarhaber) [reply content/posts/redis-assisted-client-side-caching-in-python.md]: 👍

File: content/posts/redis-assisted-client-side-caching-in-python.md (hunk 4 above)
Comment (itamarhaber) [reply content/posts/redis-assisted-client-side-caching-in-python.md]: 👍

File: content/posts/redis-assisted-client-side-caching-in-python.md (hunk 5 above)
Comment (itamarhaber) [reply content/posts/redis-assisted-client-side-caching-in-python.md]: 👍

File: content/posts/redis-assisted-client-side-caching-in-python.md (hunk 6 above)
Comment (itamarhaber) [reply content/posts/redis-assisted-client-side-caching-in-python.md]: 👍

**Pull request merged**